File: app/src/main/java/com/appttude/h_mal/days_left/FragmentList.java
package com.appttude.h_mal.days_left;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.Toast;

import com.appttude.h_mal.days_left.AddItems.AddShiftActivity;
import com.appttude.h_mal.days_left.Objects.ShiftObject;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

import static com.appttude.h_mal.days_left.Global.FirebaseClass.SHIFT_FIREBASE;
import static com.appttude.h_mal.days_left.Global.FirebaseClass.SHIFT_ID;
import static com.appttude.h_mal.days_left.Global.FirebaseClass.USER_FIREBASE;
import static com.appttude.h_mal.days_left.Global.FirebaseClass.auth;
import static com.appttude.h_mal.days_left.Global.FirebaseClass.mDatabase;


public class FragmentList extends Fragment {

    private static String TAG = "FragmentList";

    private DatabaseReference reference;
    private ListView listView;
    private FireAdapter adapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);

        reference = mDatabase.child(USER_FIREBASE).child(auth.getUid()).child(SHIFT_FIREBASE);
        reference.keepSynced(true);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_list, container, false);

        listView = view.findViewById(R.id.page_two_list);

        adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, reference);

        listView.setAdapter(adapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                String refId = adapter.getId(position);
                Intent intent = new Intent(getActivity(), AddShiftActivity.class);
                intent.putExtra(SHIFT_ID,refId);
                startActivity(intent);
            }
        });

        listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, final int position, long id) {
                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                builder.setTitle("Are you sure you want to delete?");
                builder.setNegativeButton(android.R.string.no, null);
                builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        //todo: delete from previously used if last entry
                        adapter.getRef(position).removeValue();
                    }
                });
                builder.create().show();

                return false;
            }
        });

        return view;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {

        inflater.inflate(R.menu.menu_list_fragment, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {

            case R.id.app_bar_filter:
                filterData();
                return false;

            case R.id.app_bar_soft:
                sortData();
                return false;


            default:
                break;
        }

        return false;
    }

    private void sortData(){
        final String[] grpname = {"Name","Date Added","Date of shift"};
        int checkedItem = -1;

        AlertDialog.Builder alt_bld = new AlertDialog.Builder(getContext());
        alt_bld.setTitle("Sort by:");
        alt_bld.setSingleChoiceItems(grpname, checkedItem, new DialogInterface
                .OnClickListener() {
            public void onClick(DialogInterface dialog, final int item) {
                switch (item){
                    case 0:
                        Query q1 = reference.orderByChild("abnObject/companyName").equalTo("GREEN CLOUD NURSERY");
//                                .startAt("2018-04-12").endAt("2019-03-29");

                        adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, q1);
                        break;
                    case 1:
                        adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, reference.orderByChild("dateTimeAdded"));
                        break;
                    case 2:
                        adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, reference.orderByChild("shiftDate"));
                        break;
                    default:

                }
                listView.setAdapter(adapter);
                dialog.dismiss();
            }
        });
        AlertDialog alert = alt_bld.create();
        alert.show();
    }

    private void filterData(){
        final String[] groupName = {"Name","Date Added","Shift Type"};
        int checkedItem = -1;

        final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle("Filter by:");
        builder.setSingleChoiceItems(groupName, checkedItem, new DialogInterface
                .OnClickListener() {
            public void onClick(DialogInterface dialog, final int item) {
                dialog.dismiss();

                switch (item){
                    case 0:
                        final FilterDialogClass filterDialog = new FilterDialogClass(getContext(), 0, new FilterDialogClass.FilterDialogClickListener() {
                            @Override
                            public void applyDates(String arg1, String arg2) {

                            }

                            @Override
                            public void applyAbn(String abn) {
                                applyFilter(abn,null);
                            }
                        });
                        filterDialog.show();

                        break;
                    case 1:
                        final FilterDialogClass filterDialog2 = new FilterDialogClass(getContext(), 1, new FilterDialogClass.FilterDialogClickListener() {
                            @Override
                            public void applyDates(String arg1, String arg2) {
                                applyFilter(arg1, arg2);
                            }

                            @Override
                            public void applyAbn(String abn) {

                            }
                        });
                        filterDialog2.show();
                        break;
                    case 2:
                        AlertDialog.Builder typeDialog = new AlertDialog.Builder(getContext());
                        final String[] typeString = {"Hourly", "Piece Rate"};

                        typeDialog.setSingleChoiceItems(new String[]{"Hourly", "Piece Rate"}, -1, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                Query q1 = reference.orderByChild("taskObject/workType").equalTo(typeString[which]);

                                adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, q1);
                                listView.setAdapter(adapter);
                            }
                        });
                        typeDialog.create().show();
                        break;
                    default:

                }

            }
        });
        AlertDialog alert = builder.create();
        alert.show();
    }

    private void applyFilter(String arg1, String arg2){
        Query q1;
        if (arg2 == null){
            q1 = reference.orderByChild("abnObject/abn").equalTo(arg1);
        }else {
            q1 = reference.orderByChild("shiftDate").startAt(arg1).endAt(arg2);
        }

        adapter = new FireAdapter(getActivity(), ShiftObject.class, R.layout.list_item, q1);
        listView.setAdapter(adapter);
    }

    public static Date convertDate(String s){
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
        Date d = null;
        try {
            d = sdf.parse(s);
        } catch (ParseException ex) {
            Log.v("Exception", ex.getLocalizedMessage());
        }
        finally {
            if (d == null){
                sdf.applyPattern("dd/MM/yyyy");

                try {
                    d = sdf.parse(s);
                } catch (ParseException e) {
                    e.printStackTrace();
                }
            }
        }

        return d;
    }


}

File: app/src/main/java/com/appttude/h_mal/days_left/ui/main/list/FragmentList.kt
package com.appttude.h_mal.days_left.ui.main.list

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import com.appttude.h_mal.days_left.R
import com.appttude.h_mal.days_left.ui.addShift.AddNewShiftActivity
import com.appttude.h_mal.days_left.ui.addShift.AddShiftActivity
import com.appttude.h_mal.days_left.ui.addShift.AddShiftFragment
import com.appttude.h_mal.days_left.ui.login.FullscreenActivity
import com.appttude.h_mal.days_left.ui.main.ShiftsViewModel
import com.appttude.h_mal.days_left.ui.main.list.ShiftListRecyclerAdapter.Sortation
import com.appttude.h_mal.days_left.ui.splashScreen.SplashFragment
import com.appttude.h_mal.days_left.utils.navigateTo
import com.appttude.h_mal.days_left.utils.showToast
import kotlinx.android.synthetic.main.fragment_list.*


class FragmentList : androidx.fragment.app.Fragment(), SearchView.OnQueryTextListener {
    val viewModel: ShiftsViewModel by activityViewModels()

    lateinit var fireAdapter: ShiftListRecyclerAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_list, container, false);
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //set custom firebase adapter on listview

        add_test.setOnClickListener {
            view.navigateTo(R.id.list_to_addShiftActivity)
        }

        fireAdapter =
            ShiftListRecyclerAdapter(
                view,
                viewModel
            )

        recycler_view.apply {
            adapter = fireAdapter
//            onItemLongClickListener =
//                AdapterView.OnItemLongClickListener { _, _, position, _ ->
//                    AlertDialog.Builder(context).apply {
//                        setTitle("Are you sure you want to delete?")
//                        setNegativeButton(android.R.string.no, null)
//                        setPositiveButton(android.R.string.yes) { _, _ ->
//                            fireAdapter.getRef(position).removeValue()
//                        }
//                        create().show()
//                    }
//                    true
//                }
        }
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_list_fragment, menu)

        val menuItem = menu.findItem(R.id.app_bar_filter)
        val searchView = menuItem.actionView as SearchView
        searchView.setOnQueryTextListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.app_bar_soft -> {
                sortData()
                return false
            }
        }
        return false
    }

    private fun sortData() {
        val groupName = arrayOf("Name", "Date Added", "Date of shift")
        val checkedItem = -1

        AlertDialog.Builder(context).apply {
            setTitle("Sort by:")
            setSingleChoiceItems(groupName, checkedItem) { dialog, item ->
                when (item) {
                    0 -> {
                        fireAdapter.filterShifts(Sortation.Name())
                    }
                    1 -> {
                        fireAdapter.filterShifts(Sortation.DateAdded())
                    }
                    2 ->{
                        fireAdapter.filterShifts(Sortation.ShiftDate())
                    }
                }
                dialog.dismiss()
            }
            create().show()
        }

    }

    override fun onQueryTextSubmit(query: String?): Boolean {

        return true
    }

    override fun onQueryTextChange(newText: String?): Boolean {
        if (newText.isNullOrBlank()){
            fireAdapter.filterList(null)
        }else{
            fireAdapter.filterList(newText)
        }
        return false
    }

//    private fun filterData() {
//        val groupName = arrayOf("Name", "Date Added", "Shift Type")
//        val checkedItem = -1
//
//        val builder = AlertDialog.Builder(context)
//        builder.setTitle("Filter by:")
//        builder.setSingleChoiceItems(
//            groupName,
//            checkedItem,
//            DialogInterface.OnClickListener { dialog, item ->
//                dialog.dismiss()
//
//                when (item) {
//                    0 -> {
//                        val dialogBuilder = AlertDialog.Builder(context)
//                        dialogBuilder.setTitle("Select Employer:")
//                        //get layout
//                        val dialogView = View.inflate(
//                            context,
//                            R.layout.dialog_previous_abns_used, null
//                        )
//                        //hide button
//                        dialogView.button_list_dialog.visibility = View.GONE
//                        //get listview
//                        val listView = dialogView.list_item_list_dialog
//                        //get unique abn objects
//                        val uniqueAbnObjects = viewModel.shifts
//                        //populate list in view
//                        listView.adapter =
//                            AbnListAdapter(
//                                requireContext(),
//                                uniqueAbnObjects
//                            )
//                        //on item click listener
//                        listView.setOnItemClickListener(AdapterView.OnItemClickListener { parent, view, position, id ->
//                            applyFilter(uniqueAbnObjects[position].abn!!, null)
//                        })
//                        //set view on dialog
//                        dialogBuilder.setView(dialogView)
//
//                        dialogBuilder.create().show()
//
//                    }
//                    1 -> {
//                        val customDialog =
//                            CustomDialog(requireContext())
//
//                        customDialog.setButton(BUTTON_POSITIVE,
//                            getContext()?.getString(android.R.string.yes),
//                            DialogInterface.OnClickListener { dialogNew, which ->
//                                //interface results back
//                                if (dateSelectionFrom != dateSelectionTo) {
//                                    applyFilter(dateSelectionFrom, dateSelectionTo)
//                                }
//
//                                customDialog.dismiss()
//                            })
//
//                        customDialog.create()
//                    }
//                    2 -> {
//                        val typeDialog = AlertDialog.Builder(context)
//                        val typeString = arrayOf("Hourly", "Piece Rate")
//
//                        typeDialog.setSingleChoiceItems(
//                            arrayOf("Hourly", "Piece Rate"), -1
//                        ) { dialog, which ->
//                            val q1 =
//                                ref.orderByChild("taskObject/workType").equalTo(typeString[which])
//
//                            fireAdapter = FireAdapter(
//                                activity,
//                                ShiftObject::class.java,
//                                R.layout.list_item,
//                                q1
//                            )
//                            page_two_list.adapter = fireAdapter
//                        }
//                        typeDialog.create().show()
//                    }
//                }
//            })
//    }
//
//    fun turnToUniqueAbnObject(shifts: ArrayList<ShiftObject>): List<AbnObject> {
//        val abnList = mutableListOf<AbnObject>()
//
//        shifts.forEach { shiftObject ->
//            shiftObject.abnObject?.let { abnList.add(it) }
//        }
//
//        return abnList.distinct()
//    }
//
//    fun applyFilter(arg1: String, arg2: String?) {
//        val q1: Query
//        if (arg2 == null) {
//            q1 = ref.orderByChild("abnObject/abn").equalTo(arg1)
//        } else {
//            q1 = ref.orderByChild("shiftDate").startAt(arg1).endAt(arg2)
//        }
//
//        fireAdapter = FireAdapter(
//            activity,
//            ShiftObject::class.java,
//            R.layout.list_item,
//            q1
//        )
//        page_two_list.adapter = fireAdapter
//    }


}
